doc/widget_server.md

```diff
@@ -0,0 +1,417 @@
+This is compatible with lean `v3.18.x` onwards.
+
+## Useful reference sourcefiles:
+
+- https://github.com/leanprover/lean-client-js/blob/master/lean-client-js-core/src/commands.ts
+- https://github.com/leanprover-community/lean/blob/master/src/shell/server.cpp
+- https://github.com/leanprover/vscode-lean/blob/master/infoview/widget.tsx
+- https://github.com/leanprover-community/lean/blob/master/library/init/meta/widget/basic.lean
+
+# Server protocol
+
+The `lean` executable can be started in a server mode. To do this, pass the `--server` option.
+The following arguments are also available:
+
+- `--memory`, `-M` set the maximum amount of memory that should be used by Lean (in megabytes)
+- `--timeout`, `-T` maximum number of memory allocations per task. This is a deterministic way of interrupting long running tasks.
+- `--threads`, `-j`   number of threads used to process lean files
+- `--server=file` start lean in server mode, redirecting standard input from the specified file (for debugging)
+- `--no-widgets`, `-W` turn off reporting on widgets
+
+
+Incoming messages to the server are called __requests__ and outgoing messages are called __responses__.
+All messages are formatted as JSON objects.
+Requests and responses are sent via standard input and standard output.  Each request or response is sent on a separate line.
+
+```ts
+interface Request {
+    /** command identifier string */
+    command : string;
+    seq_num? : number;
+}
+
+```
+
+`command` is one of a fixed set of command strings shown below, `seq_num` is an optional parameter which lean will use to tag a subsequent response to this request. In this way, multiple requests can be in-flight at the same time.
+
+The response from a command will implement either `CommandResponse` in the case of success, or `ErrorResponse` if an error occurred while handling the request.
+
+```ts
+interface CommandResponse extends Response {
+    response: 'ok';
+    seq_num: number;
+}
+
+interface ErrorResponse extends Response {
+    response: 'error';
+    seq_num?: number;
+    message: string;
+}
+```
+
+# Info
+
+The client may make an `info` request, this is typically triggered by the user moving their cursor to a particular point in the sourcefile.
+
+```ts
+interface InfoRequest extends Request {
+    command: 'info';
+    file_name: string;
+    line: number;
+    column: number;
+}
+```
+
+This will cause the lean server to produce the following `InfoResponse`.
+
+```ts
+interface InfoResponse extends CommandResponse {
+  record?: {
+    'full-id'?: string;
+    text?: string;
+    type?: string;
+    doc?: string;
+    source?: InfoSource;
+    tactic_params?: string[];
+    state?: GoalState;
+    widget?: WidgetIdentifier;
+  }
+}
+```
+
+In the case that the lean file is still compiling (as indicated by the orange gutter bars in vscode), then instead this will merely return a `{response: "ok"}` response.
```

Comment:
```suggestion
In the case that the Lean file is still compiling (as indicated by the orange gutter bars in VS Code), then instead this will merely return a `{response: "ok"}` response.
```